File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void gen(int i, vector<int> curr, vector<vector<int>>& ans, vector<int> arr) {
        if(i == arr.size()) {
            ans.push_back(curr);
            return;
        }
        curr.push_back(arr[i]);
        gen(i+1, curr, ans, arr);
        curr.pop_back();
        gen(i+1, curr, ans, arr);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> curr;
        vector<vector<int>> ans;
        gen(0,curr, ans, nums);
        return ans;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    void ps(int[] arr, int i, List<Integer> ans, List<List<Integer>> allSets) {
        if(i == arr.length) {
            allSets.add(new ArrayList<>(ans));
            return;
        }
        ans.add(arr[i]);
        ps(arr, i+1, ans, allSets);
        ans.remove(ans.size()-1);
        ps(arr, i+1, ans, allSets);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> allSets = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        ps(nums, 0, ans, allSets);
        return allSets;
    }
}